File: sdk/typescript/test/native.test.ts
import { describe, it, expect } from 'vitest';
import { resolve } from 'node:path';
import { NativeBackend } from '../src/native.js';

const FIXTURES = resolve(import.meta.dirname, '../../contract-tests/fixtures');

describe('NativeBackend', () => {
  describe('simple-store', () => {
    const backend = new NativeBackend(resolve(FIXTURES, 'simple-store'));

    it('reads a string value', async () => {
      expect(await backend.get('cluster', 'name')).toBe('noizu');
    });

    it('reads a nested string', async () => {
      expect(await backend.get('cluster', 'node_pool.instance_type')).toBe('m5.xlarge');
    });

    it('reads an integer', async () => {
      expect(await backend.get('cluster', 'port')).toBe(6443);
    });

    it('reads a boolean', async () => {
      expect(await backend.get('cluster', 'enabled')).toBe(true);
    });

    it('reads entire config as object', async () => {
      const result = await backend.get('cluster');
      expect(result).toBeTypeOf('object');
      expect(result).toHaveProperty('name');
      expect(result).toHaveProperty('node_pool');
    });

    it('returns null for missing path', async () => {
      expect(await backend.get('cluster', 'nonexistent')).toBeNull();
    });

    it('lists configs', async () => {
      const configs = await backend.listConfigs();
      expect(configs).toEqual(['cluster']);
    });
  });

  describe('nested-store', () => {
    const backend = new NativeBackend(resolve(FIXTURES, 'nested-store'));

    it('reads array element by index', async () => {
      expect(await backend.get('app', 'endpoints[0].host')).toBe('api.example.com');
    });

    it('reads with negative index', async () => {
      expect(await backend.get('app', 'endpoints[-1].host')).toBe('backup.example.com');
    });

    it('reads with wildcard', async () => {
      expect(await backend.get('app', 'endpoints[*].host')).toEqual([
        'api.example.com',
        'internal.example.com',
        'backup.example.com',
      ]);
    });

    it('reads length', async () => {
      expect(await backend.get('app', 'endpoints.length')).toBe(3);
    });

    it('reads chained brackets', async () => {
      expect(await backend.get('app', 'matrix[0][1]')).toBe(2);
    });

    it('lists configs', async () => {
      const configs = await backend.listConfigs();
      expect(configs).toEqual(['app']);
    });
  });
});

File: sdk/elixir/test/direnv_config/native_test.exs
defmodule DirenvConfig.NativeTest do
  use ExUnit.Case, async: true

  @fixtures_dir Path.expand("../../../contract-tests/fixtures", __DIR__)

  describe "get/3 with simple-store" do
    @store Path.join(@fixtures_dir, "simple-store")

    test "reads a simple string" do
      assert DirenvConfig.Native.get(@store, "cluster", "name") == {:ok, "noizu"}
    end

    test "reads nested string" do
      assert DirenvConfig.Native.get(@store, "cluster", "node_pool.instance_type") ==
               {:ok, "m5.xlarge"}
    end

    test "reads integer" do
      assert DirenvConfig.Native.get(@store, "cluster", "port") == {:ok, 6443}
    end

    test "reads boolean" do
      assert DirenvConfig.Native.get(@store, "cluster", "enabled") == {:ok, true}
    end

    test "reads entire config as map" do
      assert {:ok, data} = DirenvConfig.Native.get(@store, "cluster")
      assert is_map(data)
      assert Map.has_key?(data, "name")
      assert Map.has_key?(data, "node_pool")
    end

    test "missing key returns error" do
      assert DirenvConfig.Native.get(@store, "cluster", "nonexistent") == :error
    end
  end

  describe "get/3 with nested-store" do
    @nested_store Path.join(@fixtures_dir, "nested-store")

    test "array index" do
      assert DirenvConfig.Native.get(@nested_store, "app", "endpoints[0].host") ==
               {:ok, "api.example.com"}
    end

    test "wildcard" do
      assert {:ok, hosts} =
               DirenvConfig.Native.get(@nested_store, "app", "endpoints[*].host")

      assert hosts == ["api.example.com", "internal.example.com", "backup.example.com"]
    end

    test "length" do
      assert DirenvConfig.Native.get(@nested_store, "app", "endpoints.length") == {:ok, 3}
    end

    test "chained brackets" do
      assert DirenvConfig.Native.get(@nested_store, "app", "matrix[0][1]") == {:ok, 2}
    end
  end

  describe "list_configs/1" do
    test "simple store" do
      assert DirenvConfig.Native.list_configs(Path.join(@fixtures_dir, "simple-store")) ==
               {:ok, ["cluster"]}
    end

    test "nested store" do
      assert DirenvConfig.Native.list_configs(Path.join(@fixtures_dir, "nested-store")) ==
               {:ok, ["app"]}
    end
  end
end
